river/cluster/evostream.py

```diff
@@ -0,0 +1,424 @@
+                    cluster_sol_2[i].center[j],
+                    cluster_sol_1[i].center[j],
+                )
+        return cluster_sol_1, cluster_sol_2
+
+    @staticmethod
+    def _distance(point_a, point_b):
+        return math.sqrt(utils.math.minkowski_distance(point_a, point_b, 2))
+
+    def _find_closest_cluster_index(self, point, clusters):
+        min_distance = math.inf
+        closest_cluster_index = -1
+        for i, cluster_i in clusters.items():
+            distance = self._distance(clusters[i].center, point)
+            if distance < min_distance:
+                min_distance = distance
+                closest_cluster_index = i
+        return closest_cluster_index
+
+    def _gaussian_neighborhood(self, point_a, point_b):
+        distance = self._distance(point_a, point_b)
+        gaussian_neighborhood = math.exp(
+            -((distance * distance) / (3 * (self.radius * self.radius))) / 2
+        )
+        return gaussian_neighborhood
+
+    def _update(self, x):
+
+        self.time_stamp += 1
+
+        evolution_index = self.time_stamp % self.evolution_interval
+
+        if evolution_index == 0:
+            self._evolution_batch[self.evolution_interval - 1] = x
+        elif evolution_index == 1:
+            self._evolution_batch = {0: x}
+        else:
+            self._evolution_batch[evolution_index - 1] = x
+
+        temp_mc = evoStreamMicroCluster(x=x, last_update=self.time_stamp, weight=1)
+
+        merged_status = False
+
+        for micro_cluster in self._micro_clusters.values():
+            if self._distance(temp_mc.center, micro_cluster.center) < self.radius:
+                magnitude = self._gaussian_neighborhood(
+                    temp_mc.center, micro_cluster.center
+                )
+                for i in micro_cluster.center:
+                    micro_cluster.center[i] += magnitude * (
+                        temp_mc.center[i] - micro_cluster.center[i]
+                    )
+                micro_cluster.last_update = self.time_stamp
+                micro_cluster.weight = (
+                    micro_cluster.weight
+                    * (
+                        2
+                        ** (
+                            -self.decay_rate
+                            * (self.time_stamp - micro_cluster.last_update)
+                        )
+                    )
+                    + 1
+                )
+                merged_status = True
+
+        if not merged_status:
+            self._micro_clusters[len(self._micro_clusters)] = temp_mc
+
+    def _fitness(self, points, clusters):
+        ssq = 0
+        for i, point_i in points.items():
+            closest_cluster_index_i = self._find_closest_cluster_index(
+                point_i, clusters
+            )
+            distance = self._distance(point_i, clusters[closest_cluster_index_i].center)
+            ssq += distance * distance
+        return 1 / ssq
```

Comment:
What about a try/except block here to avoid division by zero? (This situation should not be common in practice, but who knows...)

Reply:
Try/except block added for a ZeroDivisionError!